Add Reducer capabilities by introducing a Reproducer that can be retrieved from TestOracle

kshehata opened the pull request:
Quick preview of what I'm working on. This adds the type information to return a Reproducer, but doesn't modify the `check()` method. Is this heading in the right direction? If so, I can add the Reducer changes tomorrow.

mrigger commented:
Looks good, thanks!

mrigger commented:
Currently, the tests are failing to the newly-added HSQLDB implementation.

kshehata commented:
Extended this with Reproducer support in SQLite3's NoREC Oracle and the StatementReducer to use it. PTAL and let me know if this looks good so far?

I'm trying to figure out now how to connect the two. The problem I'm facing is that Main has what is needed to make a new GlobalState for the Reducer to use, but ProviderAdapter has the Reproducer. Here are my ideas so far:

1. Have ProviderAdapter catch the AssertionError and return the reproducer if it exists, or rethrow if it doesn't.
2. Make another call to ProviderAdapter for reducing the query.
3. Make ProviderAdapter cache the Reproducer and return it in a separate call. This seems not great because that means the Reproducer is being cached twice.
4. Have ProviderAdapter clone the GlobalState. Is that possible?

Let me know what you think. I'm less familiar with this part of the design, so I'm not sure what the best choice is.

mrigger commented on src/sqlancer/StateToReproduce.java:
```diff
@@ -128,4 +128,8 @@ public OracleRunReproductionState createLocalState() {
     }
 
+    public void setStatements(List<Query<?>> statements) {
+        this.statements = statements;
```
Perhaps it might be safer to defensively copy the list?

mrigger reviewed (commented):
I'm not very sure. I think it would be reasonable to implement a `copy` method in `GlobalState`. Perhaps some care would be required to ensure that we deep-copy `StateToReproduce`. The other options also sound reasonable. Perhaps you can go ahead with the method you feel is most appropriate?

Btw, I'm boarding a long-distance flight and there might be some delays with my responses in the next few days. Sorry for that!

kshehata commented:
I've updated ProviderAdapter to catch AssertionError so it can pull down the Reproducer and return it to Main. Main then uses similar logic to the previous PR to do the reducing. I've also rebased to master. It's hidden behind a flag and doesn't change any of the actual output logs, just outputs to stdout. Is this now at a good state to merge? I'll squash before merging.

mrigger reviewed (approved):
Thanks, LGTM! I will use the GitHub user interface to update the branch and squash the commits.

mrigger commented:
Sorry, it seems that due to the new HSQLDB component, the build currently fails.

kshehata commented:
I've fixed HSQLDB and it should pass all tests now. Please go ahead with the squash and merge when possible. Thanks!

kshehata commented:
Test seem to be failing as apt-get cannot find Citus ?

`Unable to locate package postgresql-13-citus-10.1`

mrigger commented:
Yes, merging the PR despite the failure.

@halilozanakgul: could you please have a look at the Citus issue?

kshehata commented:
Fantastic, thanks!
